### PR TITLE
Add typing to data fetching

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Part of #7037

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [n/a] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/utilities/fetching.py [hunk 1]
```diff
@@ -158,13 +154,13 @@ def loader_iters(self) -> List[Iterator]:
 
     @property
     def state(self) -> Any:
-        def collect_state(iterator: Iterator):
+        def collect_state(iterator: Iterator) -> Any:
```
Comment (carmocca) [inline pytorch_lightning/utilities/fetching.py:157]: I'm not sure about the correct return type here: Would it be the one of `MergedIteratorState.state`?

File: pytorch_lightning/utilities/fetching.py [hunk 2]
```diff
@@ -173,9 +169,7 @@ def _attach_data_fetcher_fn(loader: DataLoader):
 
         apply_to_collection(self.loaders, (DataLoader, CycleIterator), _attach_data_fetcher_fn)
 
-    def __iter__(self) -> Generator[Tuple[Any, bool], None, None]:
-        if self.dataloader is None:
-            raise MisconfigurationException("The iterate hasn't been provided. HINT: Did you call setup function ?.")
+    def __iter__(self) -> "AbstractDataFetcher":
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:172]: What about annotating with `Self` here and only falling back to this for older python versions?

File: pytorch_lightning/utilities/fetching.py [hunk 3]
```diff
@@ -282,23 +288,21 @@ class InterBatchParallelDataFetcher(DataFetcher):
         batch 2:             [HtoD]                          [forward][backward]
     """
 
-    def __init__(self, *args, **kwargs) -> None:
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:291]: Just out of curiosity? Why do we go with *args and **kwargs here if we know the signature of `super().__init__`?

File: pytorch_lightning/utilities/fetching.py [hunk 4]
```diff
@@ -257,10 +259,14 @@ def _fetch_next_batch(self, iterator: Iterator) -> None:
...
 
     def move_to_device(self, batch: Any) -> Any:
-        if self.store_on_device and self.batch_to_device is not None:
+        if self.store_on_device:
             batch = self.batch_to_device(batch)
         return batch
 
+    def reset(self) -> None:
+        super().reset()
+        self.batches: List[Any] = []
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:268]: Can we add `batches: List[Any]` to the stub (like static attributes) if defined outside `__init__`? Then no need for typing here anymore :)

File: pytorch_lightning/utilities/fetching.py [hunk 5]
```diff
@@ -308,26 +312,23 @@ def wait(self) -> None:
...
 
 
-class StepFuncDataLoaderIter:
+class StepFuncDataLoaderIter(Iterator):
 
     """This class is a wrapper to keep track of dataloader iterator fetching event while left entirely to user
     control."""
 
-    def __init__(self, iterator: Iterator, data_fetcher: "AbstractDataFetcher"):
+    def __init__(self, iterator: Iterator, data_fetcher: "AbstractDataFetcher") -> None:
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:320]: Why can't we annotate the `data_fetcher` with the class directly? It's the same file, so shouldn't have cyclic import issues...

File: pytorch_lightning/utilities/fetching.py [hunk 6]
```diff
@@ -308,26 +312,23 @@ def wait(self) -> None:
...
-
     def __next__(self) -> Any:
         try:
             data = next(self.iterator)
             self.data_fetcher.fetched += 1
             return data
         except StopIteration:
             self.data_fetcher.done = True
-            raise StopIteration
+            raise
```
Comment (justusschock) [inline pytorch_lightning/utilities/fetching.py:331]: very nit: I find it more explicit and easier to read to do something like:

```python
try:
    ...
except StopIteration as e:
    ...
    raise e
```

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/utilities/fetching.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/utilities/fetching.py:291]: Just for simplicity, we could add them too and have them in both places

File: pytorch_lightning/utilities/fetching.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/utilities/fetching.py:172]: I tried this trick:

```python
try:
    from typing_extensions import Self
except ImportError:
    # workaround for Python 3.7.
    # see https://www.python.org/dev/peps/pep-0673/
    from typing import TypeVar

    Self = TypeVar("AbstractDataFetcher", bound="AbstractDataFetcher")
```

But getting:

```python
pytorch_lightning/utilities/fetching.py:43: error: String argument 1 "AbstractDataFetcher" to TypeVar(...) does not match variable name "Self"  [misc]
pytorch_lightning/utilities/fetching.py:43: error: Incompatible types in assignment (expression has type "TypeVar", variable has type "_SpecialForm")  [assignment]
pytorch_lightning/utilities/fetching.py:182: error: Variable "typing_extensions.Self" is not valid as a type  [valid-type]
pytorch_lightning/utilities/fetching.py:182: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases

```

Comment (justusschock) [reply pytorch_lightning/utilities/fetching.py:173]: I see. Unfortunately this is a very cryptic error message. So I also have no idea how to resolve it

**Review by akihironitta — APPROVED** [review]

File: pytorch_lightning/utilities/fetching.py [hunk 7]
```diff
@@ -308,26 +314,23 @@ def wait(self) -> None:
...
     def __next__(self) -> Any:
         try:
             data = next(self.iterator)
             self.data_fetcher.fetched += 1
             return data
-        except StopIteration:
+        except StopIteration as e:
             self.data_fetcher.done = True
-            raise StopIteration
+            raise e
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/utilities/fetching.py:333]: Battle of the nits: https://github.com/PyTorchLightning/pytorch-lightning/pull/11515#discussion_r789646065

Haha

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/fetching.py [hunk 8]
```diff
@@ -58,29 +58,35 @@ def fetching_function(self) -> Any:
...
+    def setup(self, dataloader: Iterable, **kwargs: Any) -> None:
         self._add_capture_metadata_collate(dataloader)
+        self._dataloader = dataloader
 
-        self.dataloader = dataloader
-        self.batch_to_device = batch_to_device
-
-        self._attach_data_fetcher()
+    @property
+    def dataloader(self) -> Iterable:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/utilities/fetching.py:84]: The order is already very messed. Better to leave it for a PR which moves everything where it should be.

File: pytorch_lightning/utilities/fetching.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/utilities/fetching.py:158]: I believe it should be Collection of `MergedIteratorState`

**Review by tchaton — APPROVED** [review]
LGTM !

**Pull request merged**